p2p/exchange.go

```diff
@@ -288,28 +288,18 @@ func (ex *Exchange[H]) request(
 		return nil, err
 	}
 
-	headers := make([]H, 0, len(responses))
-	for _, response := range responses {
-		if err = convertStatusCodeToError(response.StatusCode); err != nil {
-			return nil, err
-		}
-		var empty H
-		header := empty.New()
-		err := header.UnmarshalBinary(response.Body)
-		if err != nil {
-			return nil, err
-		}
-		err = validateChainID(ex.Params.chainID, header.(H).ChainID())
+	hdrs, err := processResponses[H](responses)
+	if err != nil {
+		return nil, err
+	}
+	for _, hdr := range hdrs {
+		// TODO(@Wondertan): There should be a unified header validation code path
+		err = validateChainID(ex.Params.chainID, hdr.ChainID())
 		if err != nil {
 			return nil, err
 		}
-		headers = append(headers, header.(H))
```

Comment:
The way we constructed H here is pretty ugly and does two unnecessary and costly type assertions in runtime. The new `header.New` func/generalized constructor now does it very efficiently.